homeassistant/components/bluetooth/__init__.py

```diff
@@ -0,0 +1,224 @@
+        for manufacturer in self.manufacturer_data:
+            if name := MANUFACTURERS.get(manufacturer):
+                return name
+        return None
+
+    @property
+    def manufacturer_id(self) -> int | None:
+        """Get the first manufacturer id."""
+        for manufacturer in self.manufacturer_data:
+            return manufacturer
+        return None
+
+
+BluetoothChange = Enum("BluetoothChange", "ADVERTISEMENT")
+BluetoothCallback = Callable[[BluetoothServiceInfo, BluetoothChange], Awaitable]
+
+
+@bind_hass
+async def async_register_callback(
+    hass: HomeAssistant,
+    callback: BluetoothCallback,
+    match_dict: None | dict[str, str] = None,
+) -> Callable[[], None]:
+    """Register to receive a callback on bluetooth change.
+
+    Returns a callback that can be used to cancel the registration.
+    """
+    manager: BluetoothManager = hass.data[DOMAIN]
+    return await manager.async_register_callback(callback, match_dict)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the bluetooth integration."""
+    # bt = await async_get_bluetooth(hass)
+    bluetooth: list[dict[str, str]] = []
+    bluetooth_discovery = BluetoothManager(
+        hass, bluetooth, BluetoothScanningMode.PASSIVE
+    )
+    await bluetooth_discovery.async_setup()
+    hass.data[DOMAIN] = bluetooth
+
+    # websocket_api.async_register_command(hass, list_devices)
+    # websocket_api.async_register_command(hass, set_devices)
+
+    return True
+
+
+class BluetoothManager:
+    """Manage Bluetooth."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        bluetooth: list[dict[str, str]],
+        scanning_mode: BluetoothScanningMode,
+    ) -> None:
+        """Init bluetooth discovery."""
+        self.hass = hass
+        self.scanning_mode = scanning_mode
+        self.bluetooth = bluetooth
+        self.scanner: HaBleakScanner | None = None
+        self._cancel_device_detected: CALLBACK_TYPE | None = None
+        self._scan_task: asyncio.Task | None = None
+        self._callbacks: list[tuple[BluetoothCallback, dict[str, str]]] = []
+
+    async def async_setup(self) -> None:
+        """Set up BT Discovery."""
+        try:
+            self.scanner = HaBleakScanner(
+                scanning_mode=SCANNING_MODE_TO_BLEAK[self.scanning_mode]
+            )
+        except BleakError as ex:
+            _LOGGER.warning(
+                "Could not create bluetooth scanner (is bluetooth present and enabled?): %s",
+                ex,
+            )
+            return
+        install_multiple_bleak_catcher(self.scanner)
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STARTED, self.async_start)
+
+    async def _async_start_scanner(self) -> None:
+        """Start scanner and wait until canceled."""
+        future: asyncio.Future[bool] = asyncio.Future()
+        assert self.scanner is not None
+        await self.scanner.start()
+        await future
```

Comment:
How can something set the future result if the future is a local variable and we don't pass it anywhere?

Reply:
It can't this was just for testing.

Reply:
c0eeaabf7acca039727b10ec7a049639149bd8e9